File: cpp/insert-into-a-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

// Time Complexity: O(nlogn) on average, but O(n) in the worst case of a
//                  degenerate tree (similar to a linked list)
// Space Complexity: O(n) since in the worst case we store the entire recursion
//                   stack at every node.

// Basic recursive solution
class Solution {
public:
    TreeNode* insertIntoBST(TreeNode* root, int val) {
        if (!root) {
            TreeNode* node = new TreeNode(val);
            return node;
        }

        if (val < root->val) {
            root->left = insertIntoBST(root->left, val);
        }
        else {
            root->right = insertIntoBST(root->right, val);
        }
        return root;
    }
};


// Iterative solution
class Solution {
public:
    TreeNode* insertIntoBST(TreeNode* root, int val) {
        if (!root) {
            TreeNode* node = new TreeNode(val);
            return node;
        }

        TreeNode* curr = root;
        TreeNode* node = new TreeNode(val);
        while (curr) {
            if (val < curr->val) {
                if (curr->left) {
                    curr = curr->left;
                }
                else {
                    curr->left = node;
                    break;
                }
            }
            else {
                if (curr->right) {
                    curr = curr->right;
                }
                else {
                    curr->right = node;
                    break;
                }
            }
        }
        return root ? root : node;
    }
};


// funky iterative solution with double pointer
class Solution {
public:
    TreeNode* insertIntoBST(TreeNode* root, int val) {
        if (!root) {
            TreeNode* node = new TreeNode(val);
            return node;
        }

        TreeNode** curr = &root;

        while (*curr) {
            if (val < (*curr)->val) {
                curr = &(*curr)->left;
            }
            else {
               curr = &(*curr)->right;
            }
        }

        *curr = new TreeNode(val);
        return root;
    }
};

File: py/insert-into-a-binary-search-tree.py
'''
Given the root node of a binary search tree (BST) and a value to be inserted
into the tree, insert the value into the BST. Return the root node of the
BST after the insertion. It is guaranteed that the new value does not exist
in the original BST.

Note that there may exist multiple valid ways for the insertion, as long
as the tree remains a BST after insertion. You can return any of them.

(See Leetcode site for example tree diagrams)
'''

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


# Time Complexity: O(nlogn) on average, but O(n) in the worst case of a
#                  degenerate tree (similar to a linked list)
# Space Complexity: O(n) since in the worst case we store the entire recursion
#                   stack at every node.

# Basic recursive solution
class Solution:
    def insertIntoBST(self, root: TreeNode, val: int) -> TreeNode:
        # If no root exists, create one
        if not root:
            return TreeNode(val)

        if val < root.val:
            root.left = self.insertIntoBST(root.left, val)
        else: # elif val > root.val:
            root.right = self.insertIntoBST(root.right, val)
        return root


# A slightly better recursive solution
class Solution:
    def insertIntoBST(self, root: TreeNode, val: int) -> TreeNode:
        if not root:
            return TreeNode(val)

        if val < root.val:
            if not root.left:
                node = TreeNode(val)
                root.left = node
            else:
                self.insertIntoBST(root.left, val)
        else:
            if not root.right:
                node = TreeNode(val)
                root.right = node
            else:
                self.insertIntoBST(root.right, val)
        return root


# Iterative solution (uses O(1) extra space)
class Solution:
    def insertIntoBST(self, root: TreeNode, val: int) -> TreeNode:
        if not root:
            return TreeNode(val)

        curr = root
        while curr:
            if val < curr.val:
                if not curr.left:
                    curr.left = TreeNode(val)
                    break
                else:
                    curr = curr.left
            else:
                if not curr.right:
                    curr.right = TreeNode(val)
                    break
                else:
                    curr = curr.right
        return root